### PR TITLE
utils: Don't reuse unique_ptr vars when converting qcow image

### LINKED ISSUE #1092: Daemon crashes when trying to launch core18
This happens after downloading and extracting the image. Some unhandled exception upon removing AppArmor policy:

```
[2019-09-26T10:39:16.220] [warning] [qemu-img] [2019-09-26T10:39:15.599] [debug] [daemon] Applying AppArmor policy: multipass.qemu-img
[2019-09-26T10:39:15.671] [debug] [daemon] Removing AppArmor policy:

#include <tunables/global>
profile multipass.qemu-img flags=(attach_disconnected) {
  #include <abstractions/base>

  capability dac_read_search,
    capability dac_override,

  # binary and its libs
  /usr/bin/qemu-img ixr,
  /{usr/,}lib/@{multiarch}/{,**/}*.so* rm,

  # CLASSIC ONLY: need to specify required libs from core snap
  /snap/core18/*/{,usr/}lib/@{multiarch}/{,**/}*.so* rm,

  # Subdirectory containing disk image(s)
  /** rwk,

  # Allow multipassd send qemu-img signals
  signal (receive) peer=unconfined,
}

terminate called after throwing an instance of 'std::runtime_error'
  what():  Failed to remove AppArmor policy
#include <tunables/global>
profile multipass.qemu-img flags=(attach_disconnected) {
  #include <abstractions/base>

  capability dac_read_search,
    capability dac_override,

  # binary and its libs
  /usr/bin/qemu-img ixr,
  /{usr/,}lib/@{multiarch}/{,**/}*.so* rm,

  # CLASSIC ONLY: need to specify required libs from core snap
  /snap/core18/*/{,usr/}lib/@{multiarch}/{,**/}*.so* rm,

  # Subdirectory containing disk image(s)
  /** rwk,

  # Allow multipassd send qemu-img signals
  signal (receive) peer=unconfined,
}
    : errno=254 ()

[2019-09-26T10:39:16.220] [error] [qemu-img] Process crashed
[2019-09-26T10:39:16.240] [debug] [daemon] Removing AppArmor policy:

#include <tunables/global>
profile multipass.qemu-img flags=(attach_disconnected) {
  #include <abstractions/base>

  capability dac_read_search,
    capability dac_override,

  # binary and its libs
  /usr/bin/qemu-img ixr,
  /{usr/,}lib/@{multiarch}/{,**/}*.so* rm,

  # CLASSIC ONLY: need to specify required libs from core snap
  /snap/core18/*/{,usr/}lib/@{multiarch}/{,**/}*.so* rm,

  # Subdirectory containing disk image(s)
  /** rwk,

  # Allow multipassd send qemu-img signals
  signal (receive) peer=unconfined,
}

terminate called after throwing an instance of 'std::runtime_error'
  what():  Failed to remove AppArmor policy
#include <tunables/global>
profile multipass.qemu-img flags=(attach_disconnected) {
  #include <abstractions/base>

  capability dac_read_search,
    capability dac_override,

  # binary and its libs
  /usr/bin/qemu-img ixr,
  /{usr/,}lib/@{multiarch}/{,**/}*.so* rm,

  # CLASSIC ONLY: need to specify required libs from core snap
  /snap/core18/*/{,usr/}lib/@{multiarch}/{,**/}*.so* rm,

  # Subdirectory containing disk image(s)
  /** rwk,

  # Allow multipassd send qemu-img signals
  signal (receive) peer=unconfined,
}
    : errno=254 ()
Aborted
```

### PR DESCRIPTION
**Pull request by townsend2010**
Also add unit tests for the convert_to_qcow_if_necessary() backend util function.

Fixes #1092

### CODE REVIEW

**Comment by Saviq** [comment]
Built [a snap](https://code.launchpad.net/~saviq/+snap/multipass-work/+build/900338) from [strict without this fix](https://code.launchpad.net/~saviq/+git/multipass/+ref/strict-pre-1458) and [another](https://code.launchpad.net/~saviq/+snap/multipass-work/+build/900339) from [strict with this fix](https://code.launchpad.net/~saviq/+git/multipass/+ref/strict-post-1458).

Pre-1458 failed with:
```
$ multipass launch core18
# ...
$ journalctl -u snap.multipass*
# ...
kwi 09 11:54:31 michal-laptop multipassd[2154981]: Failed to remove AppArmor policy 
                                                   #include <tunables/global>
                                                   profile multipass.qemu-img flags=(attach_disconnected) {
                                                     #include <abstractions/base>
                                                   
                                                     
                                                   
                                                     # binary and its libs
                                                     /snap/multipass/x1/usr/bin/qemu-img ixr,
                                                     /snap/multipass/x1/{usr/,}lib/@{multiarch}/{,**/}*.so* rm,
                                                   
                                                     # CLASSIC ONLY: need to specify required libs from core snap
                                                     /{,var/lib/snapd/}snap/core18/*/{,usr/}lib/@{multiarch}/{,**/}*.so* rm,
                                                   
                                                     # Subdirectory containing disk image(s)
                                                     /var/snap/multipass/common/** rwk,
                                                   
                                                     # Allow multipassd send qemu-img signals
                                                     signal (receive) peer=snap.multipass.multipassd,
                                                   }
                                                       : errno=254 ()
```

Post-1458 launched `core18` just fine.

File: tests/linux/test_backend_utils.cpp [hunk 1]
```diff
@@ -115,6 +156,36 @@ void test_image_resizing(const char* img, const mp::MemorySize& img_virtual_size
...
+    mock_factory_scope->register_callback([&](mpt::MockProcess* process) {
+        ASSERT_LE(++process_count, expected_final_process_count);
+        if (process_count == 1)
+        {
+            auto msg = QByteArray{qemuimg_info_output};
+
+            simulate_qemuimg_info_with_json(process, img_path, qemuimg_info_result, msg);
+        }
+        else
+            simulate_qemuimg_convert(process, img_path, expected_img_path, qemuimg_convert_result);
```
Comment (Saviq) [inline tests/linux/test_backend_utils.cpp:178]:
```suggestion
        else
        {
            simulate_qemuimg_convert(process, img_path, expected_img_path, qemuimg_convert_result);
        }
```

File: tests/linux/test_backend_utils.cpp [hunk 2]
```diff
@@ -238,3 +309,59 @@ TEST(BackendUtils, image_resizing_not_attempted_when_minimum_size_not_understood
...
+    const auto expected_img_path = "/fake/img/path.qcow2";
+    const auto qemuimg_info_output = "{\n    \"format\": \"raw\"\n}";
+    const auto qemuimg_info_result = success;
+    const auto attempt_convert = true;
+    const auto qemuimg_convert_result = failure;
+    const auto throw_msg_matcher = mp::make_optional(HasSubstr("qemu-img failed"));
+
+    test_image_conversion(img_path, expected_img_path, qemuimg_info_output, qemuimg_info_result, attempt_convert,
+                          qemuimg_convert_result, throw_msg_matcher);
+}
```
Comment (Saviq) [inline tests/linux/test_backend_utils.cpp:367]: Looks like a parameterized test to me :)

File: tests/linux/test_backend_utils.cpp (hunk 1 above)
Comment (townsend2010) [reply tests/linux/test_backend_utils.cpp:178]: I see another place that has the braces missing in the `else` block when the `if` block has braces.  I'll just fix all of those up in one commit.

File: tests/linux/test_backend_utils.cpp (hunk 2 above)
Comment (townsend2010) [reply tests/linux/test_backend_utils.cpp:367]: Yeah, quite a bit of code duplication here.  I'll see about doing that.

Comment (townsend2010) [reply tests/linux/test_backend_utils.cpp]: Ok, made the tests parameterized.

**Comment by townsend2010** [comment]
@Saviq,

I did make the changes you suggested.  Please give it once over :grin:  Thanks!

**Pull request merged**